agent/dockerclient/dockerapi/docker_client.go

```diff
@@ -1315,28 +1315,31 @@ func (dg *dockerGoClient) APIVersion() (dockerclient.DockerVersion, error) {
 }
 
 // Stats returns a channel of *types.StatsJSON entries for the container.
-func (dg *dockerGoClient) Stats(ctx context.Context, id string, inactivityTimeout time.Duration) (<-chan *types.StatsJSON, error) {
+func (dg *dockerGoClient) Stats(ctx context.Context, id string, inactivityTimeout time.Duration) (<-chan *types.StatsJSON, <-chan error, <-chan struct{}) {
 	subCtx, cancelRequest := context.WithCancel(ctx)
 
 	client, err := dg.sdkDockerClient()
 	if err != nil {
 		cancelRequest()
-		return nil, err
+		return nil, nil, nil
 	}
 
-	// Create channel to hold the stats
-	statsChnl := make(chan *types.StatsJSON)
+	statsC := make(chan *types.StatsJSON)
+	done := make(chan struct{})
```

Comment:
is this needed? seems like just having statsC and closing it at the end can work as well

Reply:
Yes this is needed because I have changed the behavior of the for-loop here: https://github.com/aws/amazon-ecs-agent/pull/2443/files#diff-543d5b6c06bcf2d177e4c83e48884a0fR117-R132

Closing the stats channel would cause <-dockerStats to continuously emit in the above loop.

Previously we were using a "magic" feature of Golang where a channel can be placed in a for-loop and it will exit once the channel is closed.

Reply:
the reason the loop was changed is because we're not just listening on one channel anymore (stats), we need to listen for stats and errors

Reply:
i think you can use `stat, ok := <-dockerStats` to check whether the channel is closed and break the loop when it is

Reply:
okay, will change to this